### ISSUE TITLE
PHP is missing a temporary folder.

### ISSUE THREAD

**Issue by xow** [danpoltawski/docker-moodle#9]
When testing file upload manually we get the error "PHP is missing a temporary folder."

Looks like we are just missing some configuration perhaps upload_tmp_dir = "/tmp/moodle"

Still investigating exactly how to fix it. Will update here if I find anything more

**Comment by xow** [comment]
Adding upload_tmp_dir to php.ini fixed the issue

**Comment by danpoltawski** [comment]
Thanks for creating the issue John. It's curious because the images are based on the official php-apache images, so I would have expected that to be in place in the upstream image. Will do a bit of investigation

**Comment by scara** [comment]
```
# docker run --rm -it php:7.1-apache php -i | grep upload_tmp_dir
upload_tmp_dir => no value => no value
```
Maybe it would be better to build a `php.ini` friendly for Moodle (https://docs.moodle.org/33/en/PHP#PHP_Settings ? ) and copy it into the image or, better i.e. more flexible for development purposes, to be mounted as read-only via `volumes` e.g.:
```
    volumes:
      -  ./path/to/moodle-php.ini:/usr/local/etc/php/conf.d/moodle-php.ini:ro
```
HTH,
Matteo

**Comment by danpoltawski** [comment]
It will be worth doing that if there are default we need to change.

http://php.net/manual/en/ini.core.php#ini.upload-tmp-dir says:

> If not specified PHP will use the system's default.

So I would assume it just uses /tmp in this case - surprised this isn't sufficient.

**Comment by scara** [comment]
Hi Dan,
you're right!
Quickly tested `php:7.1-apache` using HTML&PHP straight from http://php.net/manual/en/features.file-upload.post-method.php => no issue at all, using `/tmp`:
```
File is valid, and was successfully uploaded.
Here is some more debugging info:Array
(
    [userfile] => Array
        (
            [name] => SomeFewBytes.jpg
            [type] => image/jpeg
            [tmp_name] => /tmp/phpivAA4M
            [error] => 0
            [size] => 17989
        )
)
```
Sorry for my noise 😊.
At the end, a `php.ini` file  from Moodle Docs settings could be a **contributed improvement** to let Devs change e.g. the session handler or the maximum file size for uploads (`upload_max_filesize` and `post_max_size`).

**Comment by xow** [comment]
In my case, /tmp was there and no upload_tmp_dir was specified. Yet it didn't seem to use it by default.

**Comment by scara** [comment]
Hi @xow,
did you find it when playing around MDL-57698?
I'll give it a try using the Moodle image here, in one of the next days.

TIA,
Matteo

**Comment by xow** [comment]
Yep that's right. Just basically trying to get upload working for the restores.

Now I'm just leaving my machine running so I don't need to recreate my php.ini. I'm not familiar enough with docker to quickly patch this with a nice default php.ini.

**Comment by scara** [comment]
Hi @danpoltawski and @xow,
just run the composition (`bin/moodle-docker-compose up -d`) using `mysql` on `master` => confirmed, on adding a SCORM activity via DnD-ing a zip archive Moodle says:
```
Alert
PHP is missing a temporary folder.
```
Will look into it one of the next days.

HTH,
Matteo

P.S.: even the PHP upload setting is very limited (2MB), still thinking at the need for a proper `ini` configuration.

**Comment by xow** [comment]
This is what I ended up running each time I started the docker container

```
echo "upload_tmp_dir=\"/tmp/moodle\"
upload_max_filesize = 15M
post_max_size = 15M
max_execution_time = 300
error_log = /var/log/php_errors.log" > /usr/local/etc/php/php.ini && mkdir /tmp/moodle ; chmod 777 /tmp/moodle && /etc/init.d/apache2 restart
```

For whatever reason setting it to just /tmp didn't work and I never figured out why, I just used a new folder. Hacky for sure but allowed me to get testing

**Comment by danpoltawski** [comment]
Thanks for the hints, so the problem is the ADD command is reseting the /tmp permissions

**Comment by danpoltawski** [comment]
We've fixed this and it's also solved my head scratcher of #7, so thanks a lot for the report @xow and to @scara for working it out!

You would not believe how happy it makes me to have identified #7!
